<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید. 
</font>
</p>

In [2]:
import pandas as pd

data = pd.read_csv(r'D:\CodeProjects\VsCodePythonPro\MachinelearningClass\EX3\NBA2000-2009_data.csv')
data.head()


PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد. 
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

In [11]:
data['REB'] = data['OREB'] + data['DREB']

def has_double_double(row):
    ii = 0
    if row['PTS'] >= 10:
        ii += 1
    if row['REB'] >= 10:
        ii += 1
    if row['AST'] >= 10:
        ii += 1
    if row['STL'] >= 10:
        ii += 1
    if row['BLK'] >= 10:
        ii += 1
    return ii >= 2

data['Double_Double'] = data.apply(has_double_double, axis=1)

double_double_player = data[data['Double_Double']][['PLAYER', 'YEAR', 'PTS','AST','REB', 'STL', 'BLK']]
double_double_player
# data[data['Double_Double']]


PLAYER  YEAR   PTS   AST   REB  STL  BLK
2     Shaquille O'Neal  2000  28.7   3.7  12.7  0.6  2.8
5         Chris Webber  2000  27.1   4.2  11.1  1.3  1.7
13          Tim Duncan  2000  22.2   3.0  12.2  0.9  2.3
15       Kevin Garnett  2000  22.0   5.0  11.4  1.4  1.8
19     Antonio McDyess  2000  20.8   2.1  12.0  0.6  1.5
...                ...   ...   ...   ...   ...  ...  ...
1668     Dwight Howard  2009  18.3   1.8  13.2  0.9  2.8
1670    Gerald Wallace  2009  18.2   2.1  10.1  1.5  1.1
1672        Tim Duncan  2009  17.9   3.2  10.1  0.6  1.5
1686        Steve Nash  2009  16.5  11.0   3.3  0.5  0.1
1701       Troy Murphy  2009  14.6   2.1  10.2  1.0  0.5

[88 rows x 7 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

In [16]:
data['REB'] = data['OREB'] + data['DREB']

data['Missed_FG'] = data['FGA'] - data['FGM']
data['Missed_FT'] = data['FTA'] - data['FTM']

data['Value'] = (data['PTS'] + data['OREB'] + data['DREB'] + 
                     data['AST'] + data['STL'] + data['BLK']) - (
                     data['TOV'] + data['Missed_FG'] + data['Missed_FT'])

#we should use mean of 10
top10 = data[['PLAYER', 'Value']].sort_values(by='Value', ascending=False).head(10)

top10

PLAYER  Value
546      Kevin Garnett   33.2
1644      LeBron James   32.4
363      Kevin Garnett   32.2
738      Kevin Garnett   32.0
184         Tim Duncan   31.2
2     Shaquille O'Neal   31.0
1463      LeBron James   30.7
1468        Chris Paul   30.4
1262      LeBron James   30.3
921      Kevin Garnett   30.3

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [23]:
team_point = data.groupby(['YEAR', 'TEAM']).agg({'PTS': 'sum'}).reset_index()

max_team = team_point.loc[team_point.groupby('YEAR')['PTS'].idxmax()].reset_index(drop=True)

print(max_team)



   YEAR TEAM    PTS
0  2000  SAC  100.2
1  2001  LAL   95.9
2  2002  ORL   97.0
3  2003  DEN   96.0
4  2004  BOS  106.2
5  2005  WAS   96.8
6  2006  PHX  105.4
7  2007  DEN  110.7
8  2008  DEN   97.3
9  2009  OKC  102.5
